### PR TITLE
Cherry pick features/IEnumerable to release/6.0.2xx (Please don't checked in without approval from runtime team)

### PR DESCRIPTION
**Pull request by Cosifne**
features/IEnumerableAnalzyer is based on main.
So in order to get this targeting release/6.0.2.xx, I get this by running
git cherry-pick upstream/main..features/IEnumerableAnalyzer

(The diff commit could be found here https://github.com/dotnet/roslyn-analyzers/pull/5762)
All the merge commits are ignored

And there is another new commit https://github.com/dotnet/roslyn-analyzers/pull/5763/commits/1e8d4f65cc38c3bb3db9f70490945c3e951ee3a9 which fixes our config files

### CODE REVIEW

File: src/Utilities/Compiler/DiagnosticCategoryAndIdRanges.txt
Comment (Cosifne) [inline src/Utilities/Compiler/DiagnosticCategoryAndIdRanges.txt:15]: CA1850 is shipped in main, not in this branch.

**Comment by Cosifne** [comment]
I guess @jmarolf should be the right person to review this

**Review by jmarolf — APPROVED** [review]

**Comment by stephentoub** [comment]
This is merging the feature branch into a release branch? Was this analyzer approved?

**Comment by Cosifne** [comment]
> This is merging the feature branch into a release branch? Was this analyzer approved?

@stephentoub 
Are you asking for the approved PRs for this analyzer?
https://github.com/dotnet/roslyn-analyzers/pull/5304
https://github.com/dotnet/roslyn-analyzers/pull/5749
https://github.com/dotnet/roslyn-analyzers/pull/5754 

This PR contains all the changes from these PRs. (with the target branch changed to release/6.0.2xx, so I cherry-picked all the commit to another branch)

I have an offline discussion with @jmarolf and he mentioned this branch is going to ship with VS 17.1, and we could ship it if this is disabled by default.

**Comment by stephentoub** [comment]
> Are you asking for the approved PRs for this analyzer?

No, approving the existence of the analyzer.  We have a review process where new analyzers are proposed,  discussed,  approved/rejected, etc, separate from PRs.

**Comment by Cosifne** [comment]
> > Are you asking for the approved PRs for this analyzer?
> 
> No, approving the existence of the analyzer. We have a review process where new analyzers are proposed, discussed, approved/rejected, etc, separate from PRs.
> 
> cc: @mavasani, @jeffhandley, @terrajobst

@stephentoub Do I need to attend any meetings? Or are we just going to discuss it using GitHub issues? (I think this is the analyzer tracking issue https://github.com/dotnet/roslyn-analyzers/issues/3946)

**Comment by stephentoub** [comment]
> Do I need to attend any meetings? Or are we just going to discuss it using GitHub issues? (I think this is the analyzer tracking issue #3946)

I'll let cc: @mavasani, @jeffhandley, @terrajobst comment.  We may have some holes in our process here still :-) Thanks for your efforts and bearing with us.

**Comment by mavasani** [comment]
We generally follow the below guidelines:
1. Any analyzer proposal related to .NET API usage always needs to have an issue opened and triaged in the runtime repo.
2. Analyzer proposal related to generic code quality improvements doesn’t need to go through runtime triage, unless requested by any Roslyn or runtime dev. However, it does need to be discussed at least with the Roslyn team before going ahead with the implementation.

This analyzer falls under the later bucket and was discussed within the Roslyn team and deemed to be a good performance analyzer. However, given that the analyzer has bunch of hard coded logic about which runtime APIs perform an enumeration or not, it seems reasonable to open an issue on runtime repo with the analyzer proposal and design and take it through triage to ensure we have a sound implementation.


Additionally, I agree that it seems a bit too aggressive to merge a new feature into release branch, especially given this is going to be disabled by default. We should let it go through runtime triage and follow regular code flow into .NET7, even if it means we can’t enable the analyzer in Dev17.1

**Comment by Cosifne** [comment]
Thanks for Manish's comments. 
For everyone's reference, here are all the 'hard coded logic about which runtime APIs' in in this analyzer.
Basically they are used to indicate will a method enumerated the parameter or not.

https://github.com/dotnet/roslyn-analyzers/blob/927495b3992d289ba63be8c62d321954e53b8412/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs#L53

https://github.com/dotnet/roslyn-analyzers/blob/927495b3992d289ba63be8c62d321954e53b8412/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs#L87

https://github.com/dotnet/roslyn-analyzers/blob/927495b3992d289ba63be8c62d321954e53b8412/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs#L127

https://github.com/dotnet/roslyn-analyzers/blob/927495b3992d289ba63be8c62d321954e53b8412/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs#L42

Also there is another scenario that mentioned in the original PR comments https://github.com/dotnet/roslyn-analyzers/pull/5304#issuecomment-887909817
From the discussion this might be too aggressive. I can delete this part of code if needed.

**Review by jmarolf — CHANGES_REQUESTED** [review]
waiting for the runtime team to give to go ahead

**Comment by jeffhandley** [comment]
@buyaa-n -- can you extract from this an API proposal issue in dotnet/runtime to present for an API Review in January?

**Comment by buyaa-n** [comment]
> @buyaa-n -- can you extract from this an API proposal issue in dotnet/runtime to present for an API Review in January?

From the email @mavasani sent about this PR, I thought @Cosifne is preparing a proposal and interested in participating in the API review. Anyway, sure I can help @Cosifne with creating an API proposal and get it through review.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs:15]: @mavasani is this the correct namespace for this analyzer? I thought we ship in `Microsoft.NetCore.Analyzers`

Comment (Cosifne) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs:15]: Might also be good to @jmarolf for the shipping question.
I think release/6.0.2xx will be inserted into VS, and also inserted into the SDK, is that correct?

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: I wonder why we are not warning here, i believe it is Linq (IEnumerable) method not Array

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: For me this 2 calls will be difficult to implement within a one enumeration, i think its better to not flag such scenario (whenever possible)

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: Isn't it only return the `First` element in the sequesnce? I would not warn for the parameterless overload of `First()`

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: (Why the comment shown as outdated)
`i` is an array here, is that correct?

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: Emm, interesting, so maybe for methods like `except` and 'intersect' should be treated differently. Let me see if I can find a full list of methods like this

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: That's interesting. Then do we also want to exclude `ElementAt(0)` ?

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: Normally we specify the ReferenceAssemblies in the test settings instead
https://github.com/dotnet/roslyn-analyzers/blob/107d82a590a1c832f048d4f4bf1df61fd2fb3ec8/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseIntegralOrStringArgumentForIndexersTests.cs#L208

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: You can test APIs in .Net 6 https://github.com/dotnet/roslyn-analyzers/blob/107d82a590a1c832f048d4f4bf1df61fd2fb3ec8/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.Misc.cs#L21

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: > (Why the comment shown as outdated)

hm strange all of them shows outdated, not sure why

> i is an array here, is that correct?

Yes but `Array` doesn't have `Count()` method right?

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: Maybe, in case it can be excluded easily

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: Yes, my understanding it since `i` is an `Array` here, `Count()` won't call the GetEnuemrator() method to calculate the number of element, it will just fetch `Array.Count`, which is an O(1) operation. Is that correct?

Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: if you `ctrl+click` on the `Count` method of `i.Count();` above it will go to `IEnumerable.Count` extension method. As `Array` doesn't have `Count` method defined, and it implements `IEnumerable`, so when you call Array.Count it would call  `IEnumerable.Count` extension method

Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: Ah i see what you mean, even it calls `IEnumerable.Count` it would use ICollection.Cout(), never mind https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/Count.cs#L18-L21

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: I have been thinking about this for some time when editing the code.
I think this might be two kinds of  'multiple enumerations'.
1. Multiple call to the `MoveNext()` mthod, like `i.ElementAt(10)`. Calling this over `IEnumerable<T>` is the same as doing an indexing operation over a linked list.

2. The source of `IEnuemrable<T>` itself might change.
```
void Bar()
{
    IEnumerable<int> result = FetchDataFromDatabase();
    var a = result.First();
    WriteSomeDataIntoDataBase();
    var b = result.First();
}
```
In this case, even it is just called `First`, I still feel we should flag it. Because `a` and `b` might be two different values. And `First` is really querying the DB twice.

My purpose solution here would be, for a given variable which is `IEnumerable<T>`, if it is assigned from `System.Linq.Range()` or `System.Line.Repeat()`, we can safely ignore the usage of `First`, `FirstOrDefault`, etc...
Ex:
```
void Bar()
{
     var c = Enumerable.Range(1, 10000);
     c.First();
     c.FirstOrDefault();
}
```
No diagnostic should be reported.

But if the variable is assigned from some unknown method call, we should consider `First`, `FirstOrDefault`, etc...

Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: > In this case, even it is just called First, I still feel we should flag it. Because a and b might be two different values.
> And First is really querying the DB twice.

Did not understood, for me `Because a and b might be two different values.` is the reason for not flagging it, becaues the values could be different its NEED to query twice

Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: My point is this is a good case that people might misuse the Linq. (The first impression for this is that `a` should equal to `b`, but in fact because it queries DB twice, the value would be different)

Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: I am not sure how that could be a misuse of Linq, anyway i don't think we need to evaluate subsequent calls of `First()` or try to undestand the user expectations for those calls, because that doesn't seem to relate to the purpose of this analyzer (detecting multiple Enumerations), `First()` would never call `MoveNext()` more than once (by the point 1 you mentioned)

Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: OK I agree we don't add this behavior in the first revision of this analyzer.
Probably we want to wait for some feedback to see if customers really want this. 
I do really feel this might lead to bugs in reality. but we can wait to see it

Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: I am going to talk about this with our PM, because in the original post https://github.com/dotnet/roslyn-analyzers/issues/3946#issue-672955039
I think the post is showing 

> Multiple call to the MoveNext() mthod, like i.ElementAt(10). Calling this over IEnumerable<T> is the same as doing an indexing operation over a linked list.

But I want to double check if the customer may/may not want to 

> The source of IEnuemrable<T> itself might change.`

If needed I can bring that to the runtime design meeting again.

Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: I thnk you are mixing up my comment with other methods (like `ElementAt(10)`) My comment here is only about `First()` which would never make multiple call to the `MoveNext()`. For `ElementAt(10)` it would make multiple  `MoveNext()` therefore you should check/flag multiple calls

Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: No, I understand your idea here. `First()` or `FirstOrDefault()` will never call 'MoveNext()' twice. And for method like `ElementAt()` this is a different thing.

My question is mainly regarding should the analyzer report diagnostic when calling `First` or `FirstOrDefault` (the second example I posted above) This is not put in the initial post, but I want to verify with the PM about this

Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: > but I want to verify with the PM about this

Sure thing

Comment (Cosifne) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs]: Based on the offline email discussion, I will mark this as resovle.
But thank you to raise this question. : )

**Pull request closed without merging**
